### ISSUE TITLE
Inefficient compilation of x/2

### ISSUE THREAD

**Issue by Blei** [llvm-dcpu16/llvm-dcpu16#149]
The following code:

``` c
int div2(int a) {
  return a / 2;
}
```

Compiles to (using -O3):

``` dasm
:div2
    SET B, A
    SHR A, 0xf
    ADD A, B
    ASR A, 0x1
    SET PC, POP
```

A simple `ASR A, 0x1` or even `DVI A, 0x2` would be more efficient. This happens with other powers of two as well, but non-power of two divisions get just expanded to `DVI`, so I guess llvm is doing some shaky "optimisation" here.

**Comment by mgranberry** [comment]
I noticed that, but didn't dig too deeply into whether it was necessary for correctness (I suspect not).  An unsigned divide-by-two produces the expected result.  Why would it be adding in the sign bit?  Is this to ensure that rounding is consistent for positive/negative numbers?

**Comment by Blei** [comment]
Ah, you're right of course, if A would be -1 (or any negative number i suppose), `ASR` wouldn't work... Thanks.

So it does the right thing, but unfortunately this sequence of shifts and adds is slower than `DVI` on the DCPU. Maybe this optimization can be disabled somewhere?

**Comment by mgranberry** [comment]
This is being done intentionally in `TargetLowering::BuildSDIV`.  It looks to me like integer division is, in fact, cheap on this processor.  `DCPU16TargetLowering::DCPU16TargetLowering` calls `setIntDivIsCheap(false)`, but it should probably be true.  For more complex constant divides the difference is probably even more extreme.

**Comment by ghost** [comment]
Indeed, I also thought about setting the `SetIntDivIsChep()` to true. We should do that in the next update.

**Comment by ghost** [comment]
Fixed by commit 3e27d971ba408ca14b177f747a9dcee467ffddbe
